### DISCUSSION TITLE
how to dynamically add/remove submenu items?

### DISCUSSION THREAD

**Discussion by mark-summerfield** [fltk-rs/fltk-rs#1127] [Q&A]
I have a menu like this:
```
File/New
File/Open
File/Recent
File/Quit
```
The `File/Recent` has a submenu (initially empty).
I can clear the submenu using:
```
if let Some(mut item) = menubar.find_item("File/Recent\t") {
    item.clear();
   // ?????????
}
```
But what I can't figure out is how to (re)populate the item's submenu...

**Comment by MoAlyousef** [comment]
Hi
I've added MenuItem::add, add_emit, insert and insert_emit, however, MenuItems don't behave how I would expect them in menus. My current workaround to your example would be:
```rust
use fltk::{enums::*, prelude::*, *};

fn main() {
    let a = app::App::default();
    let mut win = window::Window::default().with_size(400, 300);
    let mut menubar = menu::MenuBar::new(0, 0, 400, 50, "rew");
    menubar.add("File/New\t", Shortcut::None, menu::MenuFlag::Normal, |_| {});
    menubar.add("File/Open\t", Shortcut::None, menu::MenuFlag::Normal, |_| {});
    let idx = menubar.add("File/Recent", Shortcut::None, menu::MenuFlag::Submenu, |_| {});
    menubar.add("File/Recent/First\t", Shortcut::None, menu::MenuFlag::Normal, |_| {});
    menubar.add("File/Recent/Second\t", Shortcut::None, menu::MenuFlag::Normal, |_| {});
    menubar.add("File/Quit\t", Shortcut::None, menu::MenuFlag::Normal, |_| {});
    let mut btn = button::Button::new(160, 200, 80, 30, "Modify");
    win.end();
    win.show();

    btn.set_callback(move |_| {
        menubar.clear_submenu(idx).unwrap();
        menubar.add("File/Recent/Third\t", Shortcut::None, menu::MenuFlag::Normal, |_| {});
        menubar.add("File/Recent/Fourth\t", Shortcut::None, menu::MenuFlag::Normal, |_| {});
    });

    a.run().unwrap();
}
```

**Reply by mark-summerfield** [reply]
Thanks for working that out. Do you think a "proper" solution will be found or added rather than a workaround?

I updated and tried using clear_submenu() but it fails in two ways: 1. the updates are added to a new menu item at the end of the menu so I end up with _two_ Recent Files submenus, the first empty, the second with the new items, and 2. the second menu doesn't clear so I get duplicate items.

**Reply by MoAlyousef** [reply]
Does this work:
```rust
        if let Some(mut item) = menubar.find_item("File/Recent\t") {
            item.clear();
            item.add("Recent\t/third", Shortcut::None, menu::MenuFlag::Normal, |_| {});
        }
```
Basically using add with the parent directly instead of having to pass "File/Recent\t" to add.

**Reply by MoAlyousef** [reply]
Actually clear() for MenuItem "Turns the check or radio item "off" for the menu item". I've updated the docs. And it seems FLTK doesn't have a method to remove items from a menu item.
I think MenuExt::remove, clear and clear_submenu are the only way to actually remove items from a menu.

https://www.fltk.org/doc-1.3/structFl__Menu__Item.html#a0c2aa552ddacec2b75b319fe5bc10d3e

**Reply by mark-summerfield** [reply]
OK, well as a workaround I guess I'll have to pop up a dialog listing the recent files for the user to choose from.
